rir/src/compiler/translations/rir_2_pir.cpp

```diff
@@ -0,0 +1,351 @@
+typedef rir::Opcode Opcode;
+typedef rir::BC BC;
+typedef rir::RList RList;
+
+typedef std::pair<BB*, Value*> ReturnSite;
+
+}
+
+namespace rir {
+
+pir::Function* Rir2Pir::compileFunction(SEXP function2Compile) {
+    assert(isValidClosureSEXP(function2Compile));
+    DispatchTable* tbl = DispatchTable::unpack(BODY(function2Compile));
+    auto formals = RList(FORMALS(function2Compile));
+
+    std::vector<SEXP> fmls;
+    for (auto it = formals.begin(); it != formals.end(); ++it) {
+        fmls.push_back(it.tag());
+    }
+    return this->compileFunction(tbl->first(), fmls);
+}
+
+pir::Function* Rir2Pir::compileFunction(Function* function2Compile,
+                                        std::vector<SEXP> fmls) {
+    pir::Function* pirFunction = this->compileInnerFunction(function2Compile, fmls);
+    this->optimizeFunction(pirFunction);
+    return pirFunction;
+}
+
+pir::Function* Rir2Pir::compileInnerFunction(Function* function2Compile,
+                                        std::vector<SEXP> fmls) {
+    pir::Function* pirFunction = new pir::Function(fmls);
+    IRTransformation* rir2Pir =
+        new IRTransformation(function2Compile, pirFunction);
+    this->builder = new Builder(pirFunction, Env::theContext());
+
+    this->module->functions.push_back(rir2Pir);
+    this->translateCode(function2Compile, rir2Pir->srcCode);
+    if (Verify::apply(pirFunction))
+        return pirFunction;
+    return nullptr;
+}
+
+pir::Value* Rir2Pir::translateCode(rir::Function* srcFunction, rir::Code* srcCode) {
+    this->recoverCFG(srcCode);
+    
+    std::deque<StackMachine> worklist;
+    pir::Builder* builder = this->getBuilder();
+
+    state.setPC(srcCode->code());
+    Opcode* end = srcCode->endCode();
+
+    std::vector<ReturnSite> results;
+
+    while (state.getPC() != end || !worklist.empty()) {
+        if (state.getPC() == end)
+            popFromWorklist(&worklist, builder);
+
+        BC bc = state.getCurrentBC();
+
+        if (mergepoint.count(state.getPC()) > 0) {
+            StackMachine* other = &mergepoint.at(state.getPC());
+            bool todo = state.doMerge(state.getPC(), builder, other);
+            state = mergepoint.at(state.getPC());
+            builder->next(state.getEntry());
+            if (!todo) {
+                if (worklist.empty()) {
+                    state.clear();
+                    break;
+                }
+                popFromWorklist(&worklist, builder);
+                continue;
+            }
+        }
+
+        if (bc.isJmp()) {
+            auto trg = bc.jmpTarget(state.getPC());
+            Opcode* fallpc = BC::next(state.getPC());
+            if (bc.isUncondJmp()) {
+                state.setPC(trg);
+                continue;
+            }
+
+            // Conditional jump
+            switch (bc.bc) {
+            case Opcode::brtrue_:
+            case Opcode::brfalse_: {
+                Value* v = state.pop();
+                (*builder)(new Branch(v));
```

Comment:
this is very confusing (`()` operator on a pointer).

another reason to make builder a reference, not a pointer :).

also the idea with the `operator()` is that you make the object name a verb. Like `build(...)`.

I am also fine with not using the `()` operator for this case.